node/lib/util/stitch_util.js

```diff
@@ -0,0 +1,875 @@
+        submoduleCommits: {},
+    };
+    Object.keys(subCommits).forEach(name => {
+        object.submoduleCommits[name] = subCommits[name].id().tostrS();
+    });
+    const content = JSON.stringify(object, null, 4);
+    const sig = yield ConfigUtil.defaultSignature(repo);
+    yield NodeGit.Note.create(repo,
+                              exports.referenceNoteRef,
+                              sig,
+                              sig,
+                              NodeGit.Oid.fromString(stitchedCommitSha),
+                              content,
+                              1);
+});
+
+/**
+ * Return true if the specified `newTree` and `originalTree` represent the same
+ * value and false otherwise.  `newTree` represents the same value as
+ * `originalTree` if they have the same id, or if `null === originalTree` and
+ * `newTree` has no entries.
+ *
+ * @param {NodeGit.Tree}      newTree
+ * @param {NodeGit.Tree|null} originalTree
+ * @return {Bool}
+ */
+exports.isTreeUnchanged = function (newTree, originalTree) {
+    assert.instanceOf(newTree, NodeGit.Tree);
+    if (null !== originalTree) {
+        assert.instanceOf(originalTree, NodeGit.Tree);
+    }
+    if (null === originalTree) {
+        return 0 === newTree.entries().length;
+    }
+    return newTree.id().tostrS() === originalTree.id().tostrS();
+};
+
+/**
+ * From a map containing a shas mapped to sets of direct parents, and the
+ * specified starting `entry` sha, return a list of all shas ordered from least
+ * to most dependent, that is, no sha will appear in the list before any of its
+ * ancestors.  If no relation exists between two shas, they will be ordered
+ * alphabetically.  Note that it is valid for a sha to exist as a parent from a
+ * sha in `parentMap`, however, the behavior is undefined if there are entries
+ * in 'parentMap' that are not reachable from 'entry'.
+ *
+ * @param {String} entry
+ * @param {Object} parentMap  from sha to Set of its direct parents
+ * @return {[String]}
+ */
+exports.listCommitsInOrder = function (entry, parentMap) {
+    assert.isString(entry);
+    assert.isObject(parentMap);
+
+    // First, compute the generations of the commits.  A generation '0' means
+    // that a commit has no parents.  A generation '1' means that a commit
+    // depends only on commits with 0 parents, a generation N means that a
+    // commit depends only on commits with a generation less than N.
+
+    const generations = {};
+    let queue = [entry];
+    while (0 !== queue.length) {
+        const next = queue[queue.length - 1];
+
+        // Exit if we've already computed this one; can happen if one gets into
+        // the queue more than once.
+
+        if (next in generations) {
+            queue.pop();
+            continue;                                               // CONTINUE
+        }
+        let generation = 0;
+        const parents = parentMap[next] || [];
+        for (let i = 0; i < parents.length; ++i) {
```

Comment:
Actually, there are a few for loops that should be for/of -- maybe it's worth grepping the diff  to check them all?

Reply:
Yes, I'll see if I can stomp them all. Need to get out of this habit.